### PR TITLE
Use sparse cross-correlation to accelerate local thresholding functions

### PR DESCRIPTION
**Pull request by zeroth**
## Description

correlate_sparse function computes valid cross-correlation, this function is *fast* when `kernel` is large with many zeros.

this dramatically speeds up sauvola and niblack thresholding for large kernel sizes and is equally fast for smaller ones. 

@jni developed this code in [Numba in Skan](https://github.com/jni/skan/blob/a367746c814c98124b2e5817cda2f3d97c22d9e7/skan/vendored/thresholding.py#L27) out of necessity, so this brings the skimage version into line with that.

I've added a benchmark to verify this.
- For 2d Images:
    Without sparse correlate  **780 ms**.
    With Sparse Correlate  **601ms**.

- For 3d Image:
    Without Sparse Correlate **1min 28s**.
    With Sparse Correlate  **809ms**.

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by emmanuelle** [comment]
Thank you for the PR @zeroth! A meta-question before starting a proper review: would it make sense contributing this algorithm to scipy instead? Usually the more upstream the better, if it makes sense for the upstream project of course.

**Comment by emmanuelle** [comment]
pinging @grlee77 about the interest for scipy.

**Comment by grlee77** [comment]
Yeah the current SciPy ndimage convolution code is mainly good for small, densely populated kernels. If the kernels are larger and not sparsely populated than FFT-based convolution is a reasonable alternative. The approach here does make sense for sparsely populated medium to large kernels. 

I guess the question regarding whether this should go into scipy.ndimage rather than here would be whether it is sufficiently broad in application? I am not sure how common such filters are in practice. I do appreciate you pointing out the use cases here for the niblack and sauvola thresholding methods.  Here is another publication I had seen with [sets of nd filters that each have only two non-zero entries](https://arxiv.org/abs/1708.08421). If we can make an argument that this is of general interest, then SciPy could be a reasonable fit. 

I do like the relative simplicity of the Cython code here as opposed to most of the C code underlying scipy.ndimage.

I had actually recently implemented this same sort of thing in a PR to CuPy, but we ended up not merging it (cupy/cupy#2842). This is in part because such a feature was not currently in SciPy and CuPy tends to stick closely to the NumPy and SciPy APIs.  It was also a bit slower in the fully dense case and we thought perhaps the use case for highly sparse kernels was a bit too niche to warrant supporting two parallel kernel generation codes.

File: benchmarks/benchmark_threshold_sauvola.py
Comment (jni) [inline benchmarks/benchmark_threshold_sauvola.py:10]: This shape is a Matlab-ism. =P Prefer (30, 300, 300) in NumPy. =)

File: benchmarks/benchmark_threshold_sauvola.py
Comment (jni) [inline benchmarks/benchmark_threshold_sauvola.py:19]: If you are going to use 255 as the value, make sure that `np.zeros` is called with `dtype=np.uint8`. Otherwise, use 1.0 here. See https://scikit-image.org/docs/dev/user_guide/data_types.html for details.

File: skimage/filters/_sparse.py
Comment (jni) [inline skimage/filters/_sparse.py:5]: This definitely needs a docstring with examples. I wrote this and reading the code I can't figure out what it does. 😂 There's a lesson here that everyone needs to learn multiple times in their coding life. 😂 

iirc it's like `np.meshgrid(index='ij', sparse=True)`, but broadcasting the resulting arrays together.

File: skimage/filters/_sparse.py
Comment (jni) [inline skimage/filters/_sparse.py:33]: @scikit-image/core one Q is whether we want to take *scipy* modes as input and translate to the NumPy modes. My instinct says yes since this is meant to be a "mirror" of `ndi.correlate`.

Does anyone have a time machine so we can undo the divergence between the modes? =P

**Comment by jni** [comment]
@emmanuelle @grlee77 in an ideal world this would indeed happen upstream in ndimage, but imho that is a much bigger battle. The change here is not invasive at all and I'd like to get it in locally before contemplating an upstream port. I have literally never managed to go beyond skin-deep into the ndimage code, despite several attempts.

**Comment by zeroth** [comment]
@jni I have updated the code with your recommendation please have a look .

**Comment by grlee77** [comment]
@jni, I am in agreement about going ahead with this here and can help review it now. We can revisit for SciPy later if desired.

> I have literally never managed to go beyond skin-deep into the ndimage code, despite several attempts.

I would recommend keeping this in Cython if submitted for consideration within ndimage. Although most of ndimage is in C, `ndimage.label` is actually Cython and IIRC that was encouraged as easier to maintain.

File: skimage/filters/_sparse.py
Comment (grlee77) [reply skimage/filters/_sparse.py:21]: It does seem that we have been using the SciPy mode names within `skimage.filters`, so I agree that we should also do that here for consistency.

Unfortunately we previously made the opposite decision for the `skimage.transform` functions `warp`, `reshape` and `resize`! (`warp` actually uses a [utility to convert names from numpy.pad to ndimage equivalents](https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/skimage/transform/_geometric.py#L9) so it can `ndi.map_coordinates` in some cases).

Comment (grlee77) [reply skimage/filters/_sparse.py:21]: There are more extension modes in pad than in ndimage, but just supporting the subset corresponding to those in ndimage should be sufficient here. The three that change name from numpy->ndimage are:
pad name | ndimage name
--------------|---------------------
nearest | edge
symmetric | reflect
reflect | mirror

The two that are the same for both are `wrap` and `constant`

File: skimage/filters/_sparse.py
Comment (grlee77) [inline skimage/filters/_sparse.py:57]:
```suggestion
    corner_multi_indices = np.meshgrid(*[np.arange(i) for i in result.shape], indexing='ij')
```

Isn't this just the same as calling meshgrid alone, but with the default `sparse=False`?

Comment (grlee77) [reply skimage/filters/_sparse.py:57]: Actually, you can leave `sparse=True`. The sparse arrays can be used as input to `np.ravel_multi_index` and it will take care of the broadcasting.

Comment (grlee77) [inline skimage/filters/_sparse.py:57]: I at first thought this was similar to just calling meshgrid with `sparse=False`, but on closer inspection I see that `broadcast_arrays` creates arrays that appear dense when you look at their shape, but are not actually consuming more memory due to a trick with stride=0, so this approach is preferable. 

I think it is simpler to just pass the list of sparse arrays produced by `meshgrid` directly to `np.ravel_multi_index` and allow it to handle the broadcasting, though.

File: skimage/filters/_sparse.py
Comment (grlee77) [inline skimage/filters/_sparse.py:61]:
```suggestion
    ).ravel().astype(np.intp, copy=False)
```

avoid making a copy if possible

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:57]: oooooh I was unaware that `ravel_multi_index` would handle the broadcasting! Just verified it does! Nice! NumPy is always so full of pleasant surprises. =D

@zeroth this means you can remove `np.broadcast_arrays(*)`, ie `corner_multi_indices = np.meshgrid(...)`

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:21]: > Note that although the name reflect is present in both, the behavior is not equivalent!

I know. 🤦 Again: time machine. =P

File: skimage/filters/_sparse.py
Comment (zeroth) [reply skimage/filters/_sparse.py]: @jni Tested and updated. Please have a look.

File: skimage/filters/_sparse.py
Comment (zeroth) [reply skimage/filters/_sparse.py]: @grlee77 thanks for the tip :)

**Comment by grlee77** [comment]
I just noticed that the benchmark here is added as a new file with a name corresponding to a specific function, but all other files in that folder are named after modules. For that reason, I would move this benchmark into the existing `benchmark_filters.py` file instead.

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:38]: @sciunto `ndarray of float, shape ...`? ie I don't think we've made up our mind about whether to specify the dtype, but I think we should, it's very useful.

Comment (sciunto) [reply skimage/filters/_sparse.py:38]: I put some hope in signature annotations, but for numpy, it does not work afaik. I agree that this information is useful.

Suggestion: `ndarray, dtype float, shape (Q, R,[ ...,] S)` ?

File: skimage/filters/_sparse.py
Comment (sciunto) [reply skimage/filters/_sparse.py:55]: I updated the change.

File: skimage/filters/_sparse.py
Comment (jni) [inline skimage/filters/_sparse.py:29]:
```suggestion
    mode_translation_dict = dict(nearest='edge', reflect='symmetric',
```

File: skimage/filters/_sparse.py
Comment (jni) [inline skimage/filters/_sparse.py:28]:
```suggestion
    """Convert padding modes from `ndi.correlate` to `np.pad`."""
```

File: skimage/filters/tests/test_correlate.py
Comment (jni) [inline skimage/filters/tests/test_correlate.py:8]: @zeroth could you use `skimage._shared.testing.parametrize` (example use [here](https://github.com/scikit-image/scikit-image/blob/bab6e3b92596bdf020c5a402000112b39579b06e/skimage/filters/tests/test_window.py#L17-L20)) to test all available ndimage.correlate modes? *and* additionally, test `valid` separately since that requires different handling.

Finally, you should also use `testing.raises` to test what happens when you give an even-sized kernel.

**Review by jni — APPROVED** [review]
@scikit-image/core imho this is ready to merge! 🎉 Since I co-authored bits of the code I'll let someone else review + merge.

File: skimage/filters/_sparse_cy.pyx
Comment (rfezzani) [inline skimage/filters/_sparse_cy.pyx:20]:
```suggestion
    for off, val in zip(offsets, values):
```

**Review by rfezzani — COMMENTED** [review]
Thank you @zeroth, I just have a side question: can't we release the GIL in `_correlate_sparse_offsets`?

File: skimage/filters/_sparse_cy.pyx
Comment (jni) [reply skimage/filters/_sparse_cy.pyx:20]: @rfezzani good points! One should check that the above suggestion does not harm the benchmarks.

@zeroth you can release the gil by wrapping the loop code in a

```python
with nogil:
    do_something...
```

block

Comment (zeroth) [reply skimage/filters/_sparse_cy.pyx]: Hi @rfezzani and @jni,
Quick update,
if I just update the current code to  `for off, val in zip(offsets, values):`  there no performance hit.
But, if I add 
```
with nogil:
        for off, val in zip(offsets, values):
```
Then there is a compile error 
```
with nogil:
        for off, val in zip(offsets, values):
                                            ^
skimage\filters\_sparse_cy.pyx:23:37: Converting to Python object not allowed without gil
```
I think making the thread-safe is priority hence I have just added `with nogil`. 
Let me know your thoughts.

File: skimage/filters/_sparse.py
Comment (rfezzani) [inline skimage/filters/_sparse.py:85]: Aren't `result` and `image` shapes equal? why not simply
```suggestion
    result = np.zeros_like(image)
```

File: skimage/filters/_sparse.py
Comment (rfezzani) [inline skimage/filters/_sparse.py:66]: Can't we accept array like inputs for `kernel`?
```suggestion
    kernel = np.asarray(kernel)
```

File: skimage/filters/_sparse.py
Comment (rfezzani) [inline skimage/filters/_sparse.py:95]:
```suggestion
    ).astype(np.intp, copy=False).reshape(-1)
```

File: skimage/filters/_sparse.py
Comment (rfezzani) [inline skimage/filters/_sparse.py:98]: If we simply need a view, `np.ravel` [doc](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html) advices to use `.reahspae(-1)`:
```suggestion
    _correlate_sparse_offsets(padded_image.reshape(-1), corner_indices,
                              offsets, values, result.reshape(-1))
```

File: skimage/filters/_sparse.py
Comment (rfezzani) [inline skimage/filters/_sparse.py:38]:
```suggestion
    This function is well suited when `kernel` is large with many zeros.
```

**Review by rfezzani — COMMENTED** [review]
Thank you for investigating the `nogil` point @zeroth.
I have an other question: do we want to support kernels with even shapes? Should we then consider using the `origin` argument like in `scipy.ndimage.correlate`?

File: skimage/filters/_sparse.py
Comment (zeroth) [reply skimage/filters/_sparse.py:66]: @rfezzani we had this discussion before and based on #4407  we decided to go for ndarray. https://github.com/scikit-image/scikit-image/pull/4912#discussion_r475174333 @jni please correct me if I am wrong.

File: skimage/filters/_sparse.py
Comment (zeroth) [reply skimage/filters/_sparse.py:85]: @rfezzani `result.shape == image.shape` that's for almost all the modes except `valid`. @jni I think we can do this unless I am missing something, what do you think?

**Comment by jni** [comment]
@rfezzani I think that can come in a later PR.

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:85]: As you say, this expression is valid in all cases, while the proposed expression won't work when `mode='valid'`. So we might as well keep the current expression?

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:66]: We can leave this decision for later. Honestly the entire ecosystem is having massive headaches with "array like" when it comes to typing, so I think I'd rather move to the future where everyone is responsible for providing actual NumPy arrays.

Again, we can revisit this decision later, I don't think it should hold up the PR.

File: skimage/filters/_sparse.py
Comment (rfezzani) [reply skimage/filters/_sparse.py:85]: Do I missed something? `valid` is not a valid mode for [ndi.correlate](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.correlate.html)

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:38]: @rfezzani I kinda like fast better. Let's be bold, then if someone complains that their implementation is faster, we can copy it. =P

**Comment by rfezzani** [comment]
> @rfezzani I think that can come in a later PR.

I agree, but for now we may reject even shape kernels or add a side note in the doc describing the behavior of the function in this case...

File: skimage/filters/_sparse.py
Comment (rfezzani) [reply skimage/filters/_sparse.py:38]: Well, I assume that no one will complain in any case :smile:.
I just feel that "fast" is not precise, that's it....

File: skimage/filters/_sparse.py
Comment (zeroth) [inline skimage/filters/_sparse.py:68]: @rfezzani afaik there is no mode name `valid` in `ndi.correlate` we have this specific mode check here.

File: skimage/filters/_sparse.py
Comment (rfezzani) [reply skimage/filters/_sparse.py:66]: > Honestly the entire ecosystem is having massive headaches with "array like" 

Let's assume then that `kernel` is an array and avoid the check. I don't see such validation else where in `skimage`...

File: skimage/filters/_sparse.py
Comment (rfezzani) [reply skimage/filters/_sparse.py:68]: Oh, I definitely missed something :sweat_smile:...
This is may be opinionated, but I vote for removing this option! It comes from [`scipy.signal.convolve`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve.html), and I personally  don't like it: a user needing the valid part only have to crop the output by him self!

Comment (jni) [reply skimage/filters/_sparse.py:68]: @rfezzani in this case it is incredibly useful because we don't want to pad the image if the user has already padded it. Without it, we *have* to pad.

**Comment by jni** [comment]
@rfezzani 

> for now we may reject even shape kernels

We do!

https://github.com/scikit-image/scikit-image/pull/4912/files#diff-a86dd6897f0acc876eda03ec11dbeebbR6-R23

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:38]: > I just feel that "fast" is not precise

Neither is well-suited. ;)

File: skimage/filters/thresholding.py
Comment (jni) [inline skimage/filters/thresholding.py:887]: @rfezzani over here 👆. If we don't have mode valid, we pad a potentially very large volume with a potentially very large pad, and compute a bunch of useless values, only to remove them at the end.

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py:85]: @rfezzani https://github.com/scikit-image/scikit-image/pull/4912/files#r479006134

File: skimage/filters/_sparse.py
Comment (jni) [reply skimage/filters/_sparse.py]: I just merged this suggestion. The problem with assuming is that the code will fail in a rather obscure way (list object has no attribute shape).

File: skimage/filters/thresholding.py
Comment (rfezzani) [reply skimage/filters/thresholding.py:887]: I see, now my next question is: is the acceleration coming from the kernel sparseness or from the `"valid"` mode? Can't we then simply use `scipy.signal.correlate` here instead of `ndi.correlate`?

**Comment by rfezzani** [comment]
> We do!

Excellent, I missed this part too, sorry :) It is fine then!

**Comment by jni** [comment]
The Travis failures don't appear to be related to this PR. Something about texlive *again*. 🤦‍♂

File: skimage/filters/thresholding.py
Comment (jni) [reply skimage/filters/thresholding.py:887]: No, it's the sparseness. For large kernels, SciPy (rightly) switches to FFT, which is much more expensive than 4 (2D) or 8 (3D) simple passes over the data.

**Comment by sciunto** [comment]
> The Travis failures don't appear to be related to this PR. Something about texlive _again_. man_facepalming

This time, it seems to be a timeout issue.

File: skimage/filters/_sparse.py
Comment (rfezzani) [reply skimage/filters/_sparse.py:38]: We may be can find a better wording then :smile:

**Comment by jni** [comment]
@scikit-image/core I think this PR is ready to merge, anyone else want to review?

File: skimage/filters/_sparse.py
Comment (stefanv) [reply skimage/filters/_sparse.py:38]: I'm good with fast for now.

File: skimage/filters/_sparse_cy.pyx
Comment (stefanv) [inline skimage/filters/_sparse_cy.pyx:22]: Not sure I believe that.

**Comment by stefanv** [comment]
Fantastic work, thank you @zeroth!

Request: please remove the 10x statement in a follow-up PR unless it is accurate.

File: skimage/filters/_sparse_cy.pyx
Comment (jni) [reply skimage/filters/_sparse_cy.pyx:22]: LOL It was 10x on the hardware in which I first tested this. (2015 Macbook.) On more modern hardware (my X1C gen6), it's a 4x speedup, which you can see for yourself (and I think did see for yourself? 😂) live in my [SciPy 2019 skan talk](https://www.youtube.com/watch?v=0pUPNMglnaE) (see minute 20:35 onwards). I don't have any reason to believe Cython and Numba would behave differently here.

The point of the comment is to prevent people from swapping the loop to the more "intuitive" order, so I think it should be kept in *some* form ("*up to* 10x speedup"?).

Comment (jni) [reply skimage/filters/_sparse_cy.pyx:22]: Performance is also dependent on the size of the dataset and whether it's 2D/3D (more cache misses in 3D).

Comment (jni) [reply skimage/filters/_sparse_cy.pyx:22]: Also for anyone who might be interested: the repo for that talk: https://github.com/jni/skan-talk-scipy-2019

Comment (stefanv) [reply skimage/filters/_sparse_cy.pyx:22]: Fine with me, thanks Juan!

**Pull request merged**
